Fixes #8210 Added analytics/metrics to help section

mrudultora opened the pull request:
## Purpose / Description
This PR adds analytics in Help section (Dialog box) on clicking any item.

## Fixes
Fixes #8210 

## Approach
As I was bit confused, I followed the approach suggested by David.
https://github.com/ankidroid/Anki-Android/issues/8210#issuecomment-800594220
**Also, there are some whitespaces but they are not intentional. I have properly formatted the classes and that has caused some extra line additions.**

## How Has This Been Tested?
Tested on real device (Redmi Note 7S and API 29). Analytics are logged in logcat and thus showing it's working fine.

## Learning (optional, can help others)
This PR changes used abstraction and some more changes.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison-1 reviewed (changes requested):
```
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/RecursivePictureMenuTest.java:127: error: no suitable constructor found for LinkItem(int,int,int)
> Task :AnkiDroid:compileDebugUnitTestJavaWithJavac
        return new LinkItem(R.string.help_item_ankidroid_manual, R.drawable.ic_manual_black_24dp, linkLocation);
               ^
    constructor LinkItem.LinkItem(int,int,int,int) is not applicable
      (actual and formal argument lists differ in length)
    constructor LinkItem.LinkItem(Parcel) is not applicable
      (actual and formal argument lists differ in length)
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/RecursivePictureMenuTest.java:138: error: incompatible types: Item[] cannot be converted to int
        return new ItemHeader(R.string.help_item_ankidroid_manual, R.drawable.ic_manual_black_24dp, items);
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -359,6 +359,7 @@ public static boolean isEnabled() {
 
     public static class Category {
         public static final String SYNC = "Sync";
+        public static final String LINK_CLICKED = "HelpDialog";
```
I'll defer the review mostly to Mike, he's much more clued up on the analytics side than I am.

But we presumably want a more general `LinkClicked` category if we only have a limited number

mrudultora replied:
I thought that, but was unsure that, **what should be the name?** So I used this one.

mikehardy replied:
I think literally "LinkClicked" - we're just trying to log that they clicked a link as a category, there is still the chance to put what the link was (up to a certain number of characters) so that's where the detail should go. Eventually this could also be used for outbound links on the changelog etc or anywhere we have them. It's just the idea they clicked an outbound link

mrudultora replied:
@david-allison-1 I can assign **"LinkClicked"** in LINK_CLICKED Category and append **HelpDialogBox** before each of these strings. Is that a good idea ? For example, _HelpDialog opened Using AnkiDroid_.

david-allison-1 replied:
That's a good idea, yes

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/HelpDialog.java:
```diff
@@ -86,18 +90,41 @@ public static DialogFragment createInstance(Context context) {
 
-        public RateAppItem(@StringRes int titleRes, @DrawableRes int iconRes) {
-            super(titleRes, iconRes);
+        private @StringRes
```
nit: try and keep the typical Java definition on on line - I think you can move`@StringRes` to the line above for a more sensible line break.

mrudultora replied:
Yup, sure ! That would be a nicer way..

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/RecursivePictureMenu.java:
```diff
@@ -198,13 +223,28 @@ public ItemHeader(@StringRes int titleString, int i, Item... children) {
+         * the analytics of the item clicked are send.
+         */
+        @Override
+        public void execute(AnkiActivity activity) {
```
Yo should be able to move this to the base class if it's unchanged between different classes

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/HelpDialog.java:
```diff
@@ -108,110 +135,170 @@ protected RateAppItem(Parcel in) {
                 return new RateAppItem(in);
             }
 
+
```
Could you fix the spacing changes

mrudultora replied:
I will manually remove them. Just don't know on what basis Android Studio formats files. I thought of that...

> Also, there are some whitespaces but they are not intentional. I have properly formatted the classes and that has caused some extra line additions.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/HelpDialog.java:
```diff
@@ -108,110 +135,170 @@ protected RateAppItem(Parcel in) {
+
+        @Override
+        public String getAnalyticsId(Context context) {
+            return context.getString(mAnalyticsRes);
```
You can probably pass this + variable down to the base class to save code repetition.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/HelpDialog.java:
```diff
@@ -108,110 +135,170 @@ protected RateAppItem(Parcel in) {
+        @Override
+        public void execute(AnkiActivity activity) {
+            onClicked(activity);
+            UsageAnalytics.sendAnalyticsEvent(UsageAnalytics.Category.LINK_CLICKED, getAnalyticsId(activity));
```
Optional - implementer's choice: I'd move all calls to analytics before calling `onClicked`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/RecursivePictureMenu.java:
```diff
@@ -142,52 +144,75 @@ protected void setMenuBreadcrumbHeader(MaterialDialog dialog) {
+
+        protected abstract void onClicked(AnkiActivity activity);
+
+        public abstract String getAnalyticsId(Context context);
```
I don't think this needs to be abstract if you're using an ID for all implementations

mrudultora replied:
But, all child classes must have this method. So I think, we must keep it abstract.

david-allison-1 replied:
I don't think they do - if we pass the ID to the base class (possibly a bad design, but I don't think the class is complex enough for it to matter), then they would all be the same implementation of returning `Context.getString`?

david-allison-1 commented on AnkiDroid/src/main/res/values/constants.xml:
```diff
@@ -203,6 +203,27 @@
     <string name="link_ankidroid_development_guide">https://github.com/ankidroid/Anki-Android/wiki/Development-Guide</string>
     <string name="link_ankidroid_faq">https://github.com/ankidroid/Anki-Android/wiki/FAQ</string>
     <string name="link_opencollective_donate">https://opencollective.com/ankidroid#section-contribute</string>
+    <!-- Strings used for sending analytics for HelpDialogBox -->
```
Make a note not to change these strings, as the comparison between versions of AnkiDroid will be useful

david-allison-1 commented on AnkiDroid/src/main/res/values/constants.xml:
```diff
@@ -203,6 +203,27 @@
     <string name="link_ankidroid_faq">https://github.com/ankidroid/Anki-Android/wiki/FAQ</string>
     <string name="link_opencollective_donate">https://opencollective.com/ankidroid#section-contribute</string>
+    <!-- Strings used for sending analytics for HelpDialogBox -->
+    <string name="opened_using_ankidroid">Opened Using AnkiDroid</string>
```
@mikehardy - unsure if these should be identifiers (`HelpUsingAnkiDroid`) to reduce the likelihood of people changing the strings. What's typical convention with analytics? 

We should add this to the dev guide/code style - would be useful

mikehardy replied:
I think they should be constants in the code. They must be alphabetically equal every time they are sent to the analytics provider, or they won't tally up - so there can be no variance. That's the reason for defining the category as a string constant as well by the way. Once you pick one - even though what you picked is arbitrary to begin with - that's how the data starts accumulating, and if you want it to be comparable over time and add up right, you must not change the values sent

david-allison-1 replied:
These are in `constants.xml`, so non-translatable.

But, I'm unsure here, do you have a proposal for a structure/class to store these in?

I'd be tempted to say a constants class somewhere close to `UsageAnalytics` to imply to developers that the values shouldn't change for any reason.

It'd be ideal for these to be somewhere that we can either lint or unit test in future, to ensure that nobody accidentally changes the value - don't have a concrete idea yet, but both a class or XML file would work

david-allison-1 reviewed (changes requested):
First pass through for review comments - looks good!

mrudultora commented:
@david-allison-1 Changes done as suggested by you. Now, only thing remaining is the Category tag.

david-allison-1 commented on AnkiDroid/src/main/res/values/constants.xml:
```diff
@@ -204,6 +204,7 @@
     <string name="link_ankidroid_faq">https://github.com/ankidroid/Anki-Android/wiki/FAQ</string>
     <string name="link_opencollective_donate">https://opencollective.com/ankidroid#section-contribute</string>
     <!-- Strings used for sending analytics for HelpDialogBox -->
+    <!-- Note: These Strings must note be changed -->
```
```suggestion
    <!-- Note: These Strings must not be changed as they are used for analytic comparisons between AnkiDroid versions -->
```

david-allison-1 replied:
If this goes in, we should add an issue to enforce this via a lint check

david-allison-1 replied:
#8287

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/RecursivePictureMenu.java:
```diff
@@ -144,56 +144,52 @@ protected void setMenuBreadcrumbHeader(MaterialDialog dialog) {
 
-        public abstract String getAnalyticsId(Context context);
+        public String getAnalyticsId(Context context) {
+            System.out.println("getAnalyticsId : " + context.getString(analyticsId));
```
* Use `Timber.d` and format strings rather than string concatenation
    * If you concatenate strings, then the code must execute before the method is called
    * If you use format strings, the method can say "Debug logging is disabled", exit early, and avoid any concatenation
* Extract `context.getString(analyticsId)` into a variable - use it twice, Android Studio has a really good refactoring shortcut for this, memorize the keybind.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/HelpDialog.java:
```diff
@@ -86,56 +90,75 @@ public static DialogFragment createInstance(Context context) {
+        */
+        @Override
+        public void execute(AnkiActivity activity) {
+            UsageAnalytics.sendAnalyticsEvent(UsageAnalytics.Category.LINK_CLICKED, getAnalyticsId(activity));
```
I believe this method can be moved to `Item` and made `public final void` as it should be the same for all subclasses

david-allison-1 reviewed (commented):
The amount of red in that last diff is fantastic -  60+ lines removed. We can do a little more, but this seems much better.

mrudultora commented:
@david-allison-1 Changes done ;-)

david-allison-1 reviewed (changes requested):
One pending change I feel could be done

Other than that, pending us coming up with a way to store the new analytics constants.

mrudultora commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/RecursivePictureMenu.java:
```diff
@@ -235,13 +256,14 @@ public void writeToParcel(Parcel dest, int flags) {
             }
         }
 
+
```
@david-allison-1 No, I have removed the comment from all the classes and just added that in parent class **Item**. And yes, I will remove these blank lines.

mrudultora commented:
@david-allison-1 On moving this constant strings to class would require changes. Because currently I am using string resources (int) and now they are strings. So, should I change the parameter in constructor to String?

david-allison-1 commented:
> @david-allison-1 On moving this constant strings to class would require changes. Because currently I was using string resources (int) and now they are strings. So, should I change the parameter in constructor to String?

Best to hold off on this change until we know what we're after

mikehardy commented:
Why is it not okay to just define the constants *in* UsageAnalytics? I would just define available categories in there and be done. They will only ever be used while calling a UsageAnalytics API, so they belong there IMHO

mrudultora commented:
@david-allison-1 As Mike suggested, I think we should put those strings in Category class in UsageAnalytics. Adding a small note, not to change these strings would be good. What are your views on it?

david-allison-1 commented:
Sure - sounds good to me

mrudultora commented:
@david-allison-1 I have done the changes. Following points are there :

• Moved the strings in **Category** class in UsageAnalytics and added a note there.
• Introduced string parameter rather than the int String Resource.
• Changes done accordingly in Unit Tests.
• Removed all the unnecessary whitespaces and comments (avoided repetition)

mikehardy reviewed (approved):
I gave this a read and it looks good to me *assuming* the one pending item from David is satisfied, so once David agrees, I'm +1 on it

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -359,6 +359,30 @@ public static boolean isEnabled() {
+        public static final String LINK_CLICKED = "LinkClicked";
+
+        /* These Strings must not be changed as they are used for analytic comparisons between AnkiDroid versions */
+        public static final String OPENED_HELPDIALOG = "Opened HelpDialogBox";
```
Could we consider moving these to `public static class Action {`? These aren't category labels.

@mikehardy

mrudultora replied:
It makes sense, but, making separate class for this ? If we are going to make separate class for **Actions** than we must remove all the hardcoded strings parameters passed as actions and put them here, and not just for help dialog.

david-allison-1 replied:
I feel I'm bikeshedding. 

Your call: 

* Leave it as it is, as it can quickly be refactored later
* `Actions` class with a TODO/follow-up PR
* `class LinkClicked`

david-allison-1 reviewed (commented):
We can still remove duplication of `execute` 

Implementer's choice WRT class structure of constants.

mrudultora commented:
@david-allison-1 Changes done. List of changes....

• Removed duplication of `execute`.
• Added **Actions** class in UsageAnalytics and added a note there.

david-allison-1 reviewed (approved):
Beautiful! Awaiting CI - I'll re-run the emulator tests, and will check with Mike for a quick re-review

mikehardy reviewed (approved):
This looks good to me, thanks!
As a thought: any time we add analytics because we are curious about something, we should make sure we have a real question, a real hypothesis, and we will take action on the results. Otherwise we are just collecting data for no purpose right?

So I'd like to see an issue logged for follow-up that is about like so:
- review help section analytics
- "...about 2 months after 2.15 releases to stable channel, examine help section analytics"
- "...we were curious about ..."
- "...we expect the analytics to show ... because ..."
- "...if what they show is different we will ...."

Does that make sense? I think it will be a gift to developers in the future because it will document your purpose to guard against forgetting why we did this, and it should serve as a general criteria for adding things like this in the future - if we don't have a clear reason to collect data that we can articulate like this, then we shouldn't spend any time on it IMHO

mikehardy commented:
One note is that I used squash-merge on this, A chain of 9 commits with messages like "fixed some things" are effectively worthless in a git history - they communicate nothing to future explorers of the tree. When authoring git commits consider that you are effectively writing a story that future people *will* read. Commit messages should be useful and descriptive, and small commits editing changes in response to feedback should be merged (in your local tree) into the commit they belong to if possible

mrudultora commented:
I will remember this. In some commits I just removed the whitespaces so I was not sure like _What to put in these commit message_ 😅.

mikehardy commented:
In that case I usually, mentally when I do this I am thinking "oh! that's good feedback, I'll make the change and test it then I'll edit the original commit to include the correction"

```
...make the change and test it...
git stash
git rebase -i HEAD~1  (<-- assuming the last commit is the one you are patching up, use the number of the correct commit to fix though if not the last one)
...set the commit from pick to edit for the one you want to fix in the commit selection window...
git stash pop
git add -A
git rebase --continue
...edit the commit message if needed but usually it is just a save and exit on the editor...
git push <your fork name> +<your branch name>
```
